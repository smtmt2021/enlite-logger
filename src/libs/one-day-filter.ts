/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { WattMeterBase } from '../interfaces';

const ONE_SECOUND = 1000;
const ONE_MINUTE = 60 * ONE_SECOUND;
const ONE_HOUR = 60 * ONE_MINUTE;
const ONE_DAY = 24 * ONE_HOUR;

function diff(w1: WattMeterBase, w2: WattMeterBase) {
  return w1.time.getTime() - w2.time.getTime();
}

function isOneDayBefore(w1: WattMeterBase, w2: WattMeterBase) {
  return diff(w1, w2) >= ONE_DAY;
}

function isOneHourBefore(w1: WattMeterBase, w2: WattMeterBase) {
  return diff(w1, w2) >= ONE_HOUR;
}

function calcWh(fifo: WattMeterBase[]) {
  let wh = 0;
  if (fifo.length > 1) {
    let predecessor = fifo[0];
    fifo.slice(1).forEach(w => {
      wh += predecessor.watt * diff(w, predecessor) / ONE_HOUR;
      predecessor = w;
    });
    wh += predecessor.watt * ONE_MINUTE / ONE_HOUR;
  }
  return wh;
}

export class OneDayFilter {
  private fifo: WattMeterBase[] = [];

  get hour(): number {
    if (this.fifo.length < 2) {
      return 0;
    }
    const latest = this.fifo[this.fifo.length - 1];
    const index = this.fifo.findIndex(v => !isOneHourBefore(latest, v));
    return index >= 0 ? calcWh(this.fifo.slice(index)) : 0;
  }

  get day(): number {
    return calcWh(this.fifo);
  }

  push(latest: WattMeterBase): void {
    const index = this.fifo.findIndex(v => !isOneDayBefore(latest, v));
    if (index >= 0) {
      this.fifo = this.fifo.slice(index);
    }
    this.fifo.push(latest);
  }
}
