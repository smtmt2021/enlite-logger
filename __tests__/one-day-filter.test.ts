/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { OneDayFilter } from '../src/libs/one-day-filter';

const ONE_SECOUND = 1000;
const ONE_MINUTE = 60 * ONE_SECOUND;
const ONE_HOUR = 60 * ONE_MINUTE;
const ONE_DAY = 24 * ONE_HOUR;

test('one-day-filter', () => {
  const filter = new OneDayFilter();
  let t = 0;
  for (; t < ONE_DAY; t += ONE_MINUTE) {
    const time = new Date(2000, 0, 1, 0, 0, 0, t);
    filter.push({
      watt: 100,
      time
    });
  }
  expect(filter.hour).toMatchSnapshot();
  expect(filter.day).toMatchSnapshot();

  filter.push({
    watt: 600,
    time: new Date(2000, 0, 1, 0, 0, 0, t)
  });
  expect(filter.hour).toMatchSnapshot();
  expect(filter.day).toMatchSnapshot();

  for (let i = 0; i < 59; ++i) {
    t += ONE_MINUTE;
    const time = new Date(2000, 0, 1, 0, 0, 0, t);
    filter.push({
      watt: 600,
      time
    });
  }
  expect(filter.hour).toMatchSnapshot();
  expect(filter.day).toMatchSnapshot();
});
