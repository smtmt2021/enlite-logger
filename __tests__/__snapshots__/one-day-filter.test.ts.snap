// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`one-day-filter 1`] = `100.00000000000007`;

exports[`one-day-filter 2`] = `2400.000000000005`;

exports[`one-day-filter 3`] = `108.3333333333334`;

exports[`one-day-filter 4`] = `2408.3333333333385`;

exports[`one-day-filter 5`] = `600`;

exports[`one-day-filter 6`] = `2900.000000000014`;
